File: ui-tests/PageObjects/LayoutPage.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Panes.EventLog;
using UiTests.PageObjects.Panes.VariableState;
using UiTests.PageObjects.Panes.CallTrace;
using UiTests.PageObjects.Panes.Editor;
using UiTests.PageObjects.Panes.Scratchpad;
using UiTests.PageObjects.Panes.Filesystem;
using UiTests.PageObjects.Panes.Terminal;
using UiTests.Configuration;
using UiTests.Utils;

namespace UiTests.PageObjects;

/// <summary>
/// Main layout page that contains all tabs and menu elements.
/// </summary>
public class LayoutPage : BasePage
{
    private List<EventLogPane> _eventLogTabs = new();
    private List<VariableStatePane> _programStateTabs = new();
    private List<EditorPane> _editorTabs = new();
    private List<ScratchpadPane> _scratchpadTabs = new();
    private List<FilesystemPane> _filesystemTabs = new();
    private List<TerminalOutputPane> _terminalTabs = new();
    private List<CallTracePane> _callTraceTabs = new();

    public LayoutPage(IPage page) : base(page) { }

    private async Task WaitForComponentAsync(string componentName, string selector)
    {
        var locator = Page.Locator(selector);
        try
        {
            await RetryHelpers.RetryAsync(async () =>
            {
                var count = await locator.CountAsync();
                if (count == 0)
                {
                    DebugLogger.Log($"LayoutPage: component '{componentName}' pending (selector='{selector}', count=0)");
                    return false;
                }

                var visible = await locator.First.IsVisibleAsync();
                DebugLogger.Log($"LayoutPage: component '{componentName}' ready (count={count}, firstVisible={visible})");
                return true;
            });
        }
        catch (TimeoutException ex)
        {
            var count = await locator.CountAsync();
            DebugLogger.Log($"LayoutPage: component '{componentName}' FAILED to load (selector='{selector}', final count={count})");
            throw new TimeoutException($"Component '{componentName}' (selector '{selector}') did not load; final count={count}.", ex);
        }
    }

    public Task WaitForFilesystemLoadedAsync() =>
        WaitForComponentAsync("filesystem", "div[id^='filesystemComponent']");

    public Task WaitForStateLoadedAsync() =>
        WaitForComponentAsync("state", "div[id^='stateComponent']");

    public Task WaitForCallTraceLoadedAsync() =>
        WaitForComponentAsync("calltrace", "div[id^='calltraceComponent']");

    public async Task WaitForEventLogLoadedAsync()
    {
        await WaitForComponentAsync("event-log", "div[id^='eventLogComponent']");

        var settings = AppSettingsAccessor.TryGetCurrent();
        var graceMs = settings?.Runner.ComponentLoad.EventLogGracePeriodMs ?? 0;
        if (graceMs > 0)
        {
            DebugLogger.Log($"LayoutPage: applying event-log grace delay of {graceMs}ms");
            await Task.Delay(graceMs);
        }

        var loadingCell = Page.Locator("div[id^='eventLogComponent'] td.dt-empty").Filter(new() { HasTextString = "Loading..." });
        try
        {
            await RetryHelpers.RetryAsync(async () =>
            {
                var count = await loadingCell.CountAsync();
                if (count == 0)
                {
                    DebugLogger.Log("LayoutPage: event-log data loaded (no loading placeholders present).");
                    return true;
                }

                DebugLogger.Log($"LayoutPage: event-log still loading (placeholder count={count}); waiting.");
                return false;
            });
        }
        catch (TimeoutException ex)
        {
            var remaining = await loadingCell.CountAsync();
            throw new TimeoutException($"Event log did not finish loading; {remaining} placeholder row(s) remained.", ex);
        }
    }

    public Task WaitForEditorLoadedAsync() =>
        WaitForComponentAsync("editor", "div[id^='editorComponent']");

    public Task WaitForScratchpadLoadedAsync() =>
        WaitForComponentAsync("scratchpad", "div[id^='scratchpadComponent']");

    public Task WaitForTerminalLoadedAsync() =>
        WaitForComponentAsync("terminal", "div[id^='terminalComponent']");

    public Task WaitForAllComponentsLoadedAsync()
    {
        DebugLogger.Log("LayoutPage: waiting for all components");
        var waits = new[]
        {
            WaitForFilesystemLoadedAsync(),
            WaitForStateLoadedAsync(),
            WaitForCallTraceLoadedAsync(),
            WaitForEventLogLoadedAsync(),
            WaitForEditorLoadedAsync(),
            WaitForTerminalLoadedAsync(),
            WaitForScratchpadLoadedAsync()
        };
        return Task.WhenAll(waits);
    }

    #region Debug Buttons
    public ILocator RunToEntryButton() => Page.Locator("#run-to-entry-debug");
    public ILocator ContinueButton() => Page.Locator("#continue-debug");
    public ILocator ReverseContinueButton() => Page.Locator("#reverse-continue-debug");
    public ILocator StepOutButton() => Page.Locator("#step-out-debug");
    public ILocator ReverseStepOutButton() => Page.Locator("#reverse-step-out-debug");
    public ILocator StepInButton() => Page.Locator("#step-in-debug");
    public ILocator ReverseStepInButton() => Page.Locator("#reverse-step-in-debug");
    public ILocator NextButton() => Page.Locator("#next-debug");
    public ILocator ReverseNextButton() => Page.Locator("#reverse-next-debug");

    public ILocator OperationStatus() => Page.Locator("#operation-status");
    public ILocator StatusBusyIndicator() => Page.Locator(".status-notification.is-active");
    #endregion

    public async Task<IReadOnlyList<EventLogPane>> EventLogTabsAsync(bool forceReload = false)
    {
        if (forceReload || _eventLogTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='eventLogComponent-']").AllAsync();
            _eventLogTabs = roots.Select(r => new EventLogPane(Page, r, "EVENT LOG")).ToList();
        }
        return _eventLogTabs;
    }

    public async Task<IReadOnlyList<VariableStatePane>> ProgramStateTabsAsync(bool forceReload = false)
    {
        if (forceReload || _programStateTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='stateComponent']").AllAsync();
            _programStateTabs = roots.Select(r => new VariableStatePane(Page, r, "STATE")).ToList();
        }
        return _programStateTabs;
    }

    public async Task<IReadOnlyList<EditorPane>> EditorTabsAsync(bool forceReload = false)
    {
        if (forceReload || _editorTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='editorComponent']").AllAsync();
            var tabs = new List<EditorPane>();
            foreach (var r in roots)
            {
                var idAttr = await r.GetAttributeAsync("id") ?? string.Empty;
                var filePath = await r.GetAttributeAsync("data-label") ?? string.Empty;
                var segments = filePath.Split('/', System.StringSplitOptions.RemoveEmptyEntries);
                var fileName = segments.LastOrDefault() ?? string.Empty;
                var tabButtonText = string.Join('/', segments.TakeLast(2));
                var idMatch = Regex.Match(idAttr, @"(\d+)");
                var idNumber = idMatch.Success ? int.Parse(idMatch.Groups[1].Value) : -1;
                var paneRoot = Page.Locator($"#{idAttr}");
                var pane = new EditorPane(Page, paneRoot, tabButtonText, idNumber, filePath, fileName);
                tabs.Add(pane);
            }
            _editorTabs = tabs;
        }
        return _editorTabs;
    }

    public async Task<IReadOnlyList<ScratchpadPane>> ScratchpadTabsAsync(bool forceReload = false)
    {
        if (forceReload || _scratchpadTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='scratchpadComponent']").AllAsync();
            _scratchpadTabs = roots
                .Select(r => new ScratchpadPane(Page, r, "SCRATCHPAD"))
                .ToList();
        }

        return _scratchpadTabs;
    }

    public async Task<IReadOnlyList<FilesystemPane>> FilesystemTabsAsync(bool forceReload = false)
    {
        if (forceReload || _filesystemTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='filesystemComponent']").AllAsync();
            _filesystemTabs = roots
                .Select(r => new FilesystemPane(Page, r, "FILES"))
                .ToList();
        }

        return _filesystemTabs;
    }

    public async Task<IReadOnlyList<TerminalOutputPane>> TerminalTabsAsync(bool forceReload = false)
    {
        if (forceReload || _terminalTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='terminalComponent']").AllAsync();
            _terminalTabs = roots
                .Select(r => new TerminalOutputPane(Page, r, "TERMINAL"))
                .ToList();
        }

        return _terminalTabs;
    }

    public async Task<IReadOnlyList<CallTracePane>> CallTraceTabsAsync(bool forceReload = false)
    {
        if (forceReload || _callTraceTabs.Count == 0)
        {
            var roots = await Page.Locator("div[id^='calltraceComponent']").AllAsync();
            _callTraceTabs = roots
                .Select(r => new CallTracePane(Page, r, "CALLTRACE"))
                .ToList();
        }

        return _callTraceTabs;
    }

    public ILocator MenuRootButton() => Page.Locator("#menu-root-name");
    public ILocator MenuSearchTextBox() => Page.Locator("#menu-search-text");

}

File: tsc-ui-tests/page-objects/layout-page.ts
import type { Locator, Page } from "@playwright/test";
import { BasePage } from "./base-page";
import { EditorPane } from "./panes/editor/editor-pane";
import { EventLogPane } from "./panes/event-log/event-log-pane";
import { CallTracePane } from "./panes/call-trace/call-trace-pane";
import { ScratchpadPane } from "./panes/scratchpad/scratchpad-pane";
import { FilesystemPane } from "./panes/filesystem/filesystem-pane";
import { TerminalOutputPane } from "./panes/terminal/terminal-output-pane";
import { VariableStatePane } from "./panes/variable-state/variable-state-pane";
import { retry } from "../lib/retry-helpers";
import { debugLogger } from "../lib/debug-logger";

const EVENT_LOG_LOADING_RETRY_ATTEMPTS = 30;

/**
 * Main layout page that contains all pane tabs, debug buttons, and menu elements.
 *
 * Port of ui-tests/PageObjects/LayoutPage.cs with all enhanced pane accessors.
 */
export class LayoutPage extends BasePage {
  private eventLogTabsCache: EventLogPane[] = [];
  private programStateTabsCache: VariableStatePane[] = [];
  private editorTabsCache: EditorPane[] = [];
  private scratchpadTabsCache: ScratchpadPane[] = [];
  private filesystemTabsCache: FilesystemPane[] = [];
  private terminalTabsCache: TerminalOutputPane[] = [];
  private callTraceTabsCache: CallTracePane[] = [];

  // ---------------------------------------------------------------------------
  // Component wait methods
  // ---------------------------------------------------------------------------

  private async waitForComponent(
    componentName: string,
    selector: string,
  ): Promise<void> {
    const locator = this.page.locator(selector);
    try {
      await retry(async () => {
        const count = await locator.count();
        if (count === 0) {
          debugLogger.log(
            `LayoutPage: component '${componentName}' pending (selector='${selector}', count=0)`,
          );
          return false;
        }
        const visible = await locator.first().isVisible();
        debugLogger.log(
          `LayoutPage: component '${componentName}' ready (count=${count}, firstVisible=${visible})`,
        );
        return true;
      });
    } catch (ex) {
      const count = await locator.count();
      const candidateIds: string[] = await this.page.evaluate(() =>
        Array.from(document.querySelectorAll("div[id]"))
          .map((el) => el.id)
          .filter((id) => id.toLowerCase().includes("component"))
          .slice(0, 25),
      );
      const bodySummary: string = await this.page.evaluate(
        () => (document.body?.innerText ?? "").slice(0, 600),
      );
      const knownComponents =
        candidateIds.length === 0 ? "none" : candidateIds.join(", ");
      debugLogger.log(
        `LayoutPage: component '${componentName}' FAILED to load (selector='${selector}', final count=${count}, known=${knownComponents}, body='${bodySummary}')`,
      );
      throw new Error(
        `Component '${componentName}' (selector '${selector}') did not load; final count=${count}. ` +
          `Known component-like ids: ${knownComponents}. Body sample: ${bodySummary}`,
        { cause: ex instanceof Error ? ex : undefined },
      );
    }
  }

  async waitForFilesystemLoaded(): Promise<void> {
    await this.waitForComponent("filesystem", "div[id^='filesystemComponent']");
  }

  async waitForStateLoaded(): Promise<void> {
    await this.waitForComponent("state", "div[id^='stateComponent']");
  }

  async waitForCallTraceLoaded(): Promise<void> {
    await this.waitForComponent("calltrace", "div[id^='calltraceComponent']");
  }

  async waitForEventLogLoaded(): Promise<void> {
    await this.waitForComponent("event-log", "div[id^='eventLogComponent']");

    // Wait for "Loading..." placeholder cells to disappear
    const loadingCell = this.page
      .locator("div[id^='eventLogComponent'] td.dt-empty")
      .filter({ hasText: "Loading..." });

    try {
      await retry(
        async () => {
          const count = await loadingCell.count();
          if (count === 0) {
            debugLogger.log(
              "LayoutPage: event-log data loaded (no loading placeholders present).",
            );
            return true;
          }
          debugLogger.log(
            `LayoutPage: event-log still loading (placeholder count=${count}); waiting.`,
          );
          return false;
        },
        { maxAttempts: EVENT_LOG_LOADING_RETRY_ATTEMPTS },
      );
    } catch (ex) {
      const remaining = await loadingCell.count();
      throw new Error(
        `Event log did not finish loading; ${remaining} placeholder row(s) remained.`,
        { cause: ex instanceof Error ? ex : undefined },
      );
    }
  }

  async waitForEditorLoaded(): Promise<void> {
    await this.waitForComponent("editor", "div[id^='editorComponent']");
  }

  async waitForScratchpadLoaded(): Promise<void> {
    await this.waitForComponent("scratchpad", "div[id^='scratchpadComponent']");
  }

  async waitForTerminalLoaded(): Promise<void> {
    await this.waitForComponent("terminal", "div[id^='terminalComponent']");
  }

  /**
   * Waits for the trace to be fully loaded by checking the document title.
   */
  async waitForTraceLoaded(): Promise<void> {
    debugLogger.log("LayoutPage: waiting for trace to be loaded (checking document title)");
    await retry(async () => {
      const title = await this.page.title();
      const isLoaded = title.toLowerCase().includes("trace");
      if (!isLoaded) {
        debugLogger.log(`LayoutPage: trace not yet loaded (title='${title}')`);
      }
      return isLoaded;
    });
    debugLogger.log("LayoutPage: trace loaded confirmed via title");
  }

  async waitForAllComponentsLoaded(): Promise<void> {
    debugLogger.log("LayoutPage: waiting for all components");
    await Promise.all([
      this.waitForFilesystemLoaded(),
      this.waitForStateLoaded(),
      this.waitForCallTraceLoaded(),
      this.waitForEventLogLoaded(),
      this.waitForEditorLoaded(),
      this.waitForTerminalLoaded(),
      this.waitForScratchpadLoaded(),
    ]);
  }

  /**
   * Waits for base components excluding the editor (which loads dynamically
   * after backend sends CtCompleteMove).
   */
  async waitForBaseComponentsLoaded(): Promise<void> {
    debugLogger.log("LayoutPage: waiting for base components (excluding editor)");
    await Promise.all([
      this.waitForFilesystemLoaded(),
      this.waitForStateLoaded(),
      this.waitForCallTraceLoaded(),
      this.waitForEventLogLoaded(),
      this.waitForTerminalLoaded(),
      this.waitForScratchpadLoaded(),
    ]);
  }

  // ---------------------------------------------------------------------------
  // Debug buttons
  // ---------------------------------------------------------------------------

  runToEntryButton(): Locator {
    return this.page.locator("#run-to-entry-debug");
  }
  continueButton(): Locator {
    return this.page.locator("#continue-debug");
  }
  reverseContinueButton(): Locator {
    return this.page.locator("#reverse-continue-debug");
  }
  stepOutButton(): Locator {
    return this.page.locator("#step-out-debug");
  }
  reverseStepOutButton(): Locator {
    return this.page.locator("#reverse-step-out-debug");
  }
  stepInButton(): Locator {
    return this.page.locator("#step-in-debug");
  }
  reverseStepInButton(): Locator {
    return this.page.locator("#reverse-step-in-debug");
  }
  nextButton(): Locator {
    return this.page.locator("#next-debug");
  }
  reverseNextButton(): Locator {
    return this.page.locator("#reverse-next-debug");
  }

  // ---------------------------------------------------------------------------
  // Status indicators
  // ---------------------------------------------------------------------------

  operationStatus(): Locator {
    return this.page.locator("#operation-status");
  }

  statusBusyIndicator(): Locator {
    return this.page.locator(".status-notification.is-active");
  }

  // ---------------------------------------------------------------------------
  // Menu elements
  // ---------------------------------------------------------------------------

  menuRootButton(): Locator {
    return this.page.locator("#menu-root-name");
  }

  menuSearchTextBox(): Locator {
    return this.page.locator("#menu-search-text");
  }

  // ---------------------------------------------------------------------------
  // Pane tab accessors (cached with forceReload)
  // ---------------------------------------------------------------------------

  async eventLogTabs(forceReload = false): Promise<EventLogPane[]> {
    if (forceReload || this.eventLogTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='eventLogComponent-']")
        .all();
      this.eventLogTabsCache = roots.map(
        (r) => new EventLogPane(this.page, r, "EVENT LOG"),
      );
    }
    return this.eventLogTabsCache;
  }

  async programStateTabs(forceReload = false): Promise<VariableStatePane[]> {
    if (forceReload || this.programStateTabsCache.length === 0) {
      const roots = await this.page.locator("div[id^='stateComponent']").all();
      this.programStateTabsCache = roots.map(
        (r) => new VariableStatePane(this.page, r, "STATE"),
      );
    }
    return this.programStateTabsCache;
  }

  async editorTabs(forceReload = false): Promise<EditorPane[]> {
    if (forceReload || this.editorTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='editorComponent']")
        .all();
      const tabs: EditorPane[] = [];
      for (const r of roots) {
        const idAttr = (await r.getAttribute("id")) ?? "";
        const filePath = (await r.getAttribute("data-label")) ?? "";
        const segments = filePath.split("/").filter(Boolean);
        const fileName = segments[segments.length - 1] ?? "";
        const tabButtonText =
          segments.length >= 2
            ? segments.slice(-2).join("/")
            : fileName;
        const idMatch = idAttr.match(/(\d+)/);
        const idNumber = idMatch ? parseInt(idMatch[1], 10) : -1;
        const paneRoot = this.page.locator(`#${idAttr}`);
        tabs.push(
          new EditorPane(
            this.page,
            paneRoot,
            tabButtonText,
            idNumber,
            filePath,
            fileName,
          ),
        );
      }
      this.editorTabsCache = tabs;
    }
    return this.editorTabsCache;
  }

  async callTraceTabs(forceReload = false): Promise<CallTracePane[]> {
    if (forceReload || this.callTraceTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='calltraceComponent']")
        .all();
      this.callTraceTabsCache = roots.map(
        (r) => new CallTracePane(this.page, r, "CALLTRACE"),
      );
    }
    return this.callTraceTabsCache;
  }

  async scratchpadTabs(forceReload = false): Promise<ScratchpadPane[]> {
    if (forceReload || this.scratchpadTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='scratchpadComponent']")
        .all();
      this.scratchpadTabsCache = roots.map(
        (r) => new ScratchpadPane(this.page, r, "SCRATCHPAD"),
      );
    }
    return this.scratchpadTabsCache;
  }

  async filesystemTabs(forceReload = false): Promise<FilesystemPane[]> {
    if (forceReload || this.filesystemTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='filesystemComponent']")
        .all();
      this.filesystemTabsCache = roots.map(
        (r) => new FilesystemPane(this.page, r, "FILES"),
      );
    }
    return this.filesystemTabsCache;
  }

  async terminalTabs(forceReload = false): Promise<TerminalOutputPane[]> {
    if (forceReload || this.terminalTabsCache.length === 0) {
      const roots = await this.page
        .locator("div[id^='terminalComponent']")
        .all();
      this.terminalTabsCache = roots.map(
        (r) => new TerminalOutputPane(this.page, r, "TERMINAL"),
      );
    }
    return this.terminalTabsCache;
  }
}
